R/module_snapshot_manager.R

```diff
@@ -144,7 +155,63 @@ snapshot_manager_srv <- function(id, slices_global, mapping_matrix, filtered_dat
       }
     })
 
-    # Restore initial state.
+    # Upload a snapshot file ----
+    # Select file.
+    observeEvent(input$snapshot_load, {
+      showModal(
+        modalDialog(
+          fileInput(ns("snapshot_file"), "Choose snapshot file", accept = ".json", width = "100%"),
+          textInput(
+            ns("snapshot_name"),
+            "Name the snapshot (optional)",
+            width = "100%",
+            placeholder = "Meaningful, unique name"
+          ),
+          footer = tagList(
+            actionButton(ns("snaphot_file_accept"), "Accept", icon = icon("thumbs-up")),
+            modalButton(label = "Cancel", icon = icon("thumbs-down"))
+          )
+        )
+      )
+    })
+    # Store new snapshot to list and restore filter states.
+    observeEvent(input$snaphot_file_accept, {
+      snapshot_name <- trimws(input$snapshot_name)
+      if (identical(snapshot_name, "")) {
+        snapshot_name <- tools::file_path_sans_ext(input$snapshot_file$name)
+      }
+      if (is.element(make.names(snapshot_name), make.names(names(snapshot_history())))) {
```

Comment:
```suggestion
      if (tolower(tools::file_ext(input$snapshot_file$name)) != "json") {
    showNotification(
      "Please select a JSON file.",
      type = "error"
    )
  } else if (is.element(make.names(snapshot_name), make.names(names(snapshot_history())))) {
```
if user add any- other format it crash the application.

Reply:
Interesting. I thought limiting the fileInput to accept only JSON files should not allow the fileInput UI to do this.
What file did you upload?

Reply:
This is covered by `fileInput(ns("snapshot_file"), "Choose snapshot file", accept = ".json", width = "100%")`

Reply:
exactly

Reply:
Indeed, the system accepts JSON as the designated format, but users can still switch to a different format while uploading file. HEre I changed it to all files and uploaded Zip. 
<img width="462" alt="image" src="https://github.com/insightsengineering/teal/assets/6700955/31bacedc-3a0b-45dd-a4aa-7b4453fef4a5">
<img width="445" alt="image" src="https://github.com/insightsengineering/teal/assets/6700955/597ffc2c-31e2-498c-8f05-c4deaed585b8">

Reply:
imo I think that would be a feature request for the `shiny::fileInput` and out of scope for us.

Reply:
Oh, by the way, the app crashes with 
```
Warning: Error in slices_restore: Assertion on 'file' failed: File extension must be in {'json'} (case insensitive), but file name is '/tmp/RtmpmTbIfP/e9f1938f5618b837dde5618a/0.txt'.
```

Reply:
In my opinion, we should incorporate an additional failsafe for users instead of allowing the app to crash. The above error is not visible to the user anyway.

Reply:
I agree that the app should not crash. I would go with a general tryCatch and show a meaningful error message (like "Error while using this file. Please make sure you upload a ....") that the user can act upon rather than us fixing for every user-based edge case that is simply outside the scope of "our" package.
This is the same case as using external API. fileUploads are beyond our control and tryCatch makes sense here.